hazelcast/src/test/java/com/hazelcast/test/ClusterFailureTestSupport.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.test;
+
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.test.TestHazelcastFactory;
+import com.hazelcast.config.Config;
+import com.hazelcast.core.HazelcastInstance;
+import java.util.Arrays;
+
+public final class ClusterFailureTestSupport {
+    private ClusterFailureTestSupport() {
+    }
+
+    public abstract static class SingleFailingInstanceClusterFailure {
+        protected HazelcastInstance[] hazelcastInstances;
+        protected HazelcastInstance failingInstance;
+        protected TestHazelcastFactory factory = new TestHazelcastFactory();
+        protected HazelcastInstance client;
+        protected int initialClusterMembersCount;
+        protected Config config;
+
+        public void initialize(int initialClusterMembersCount, Config config) {
+            this.initialClusterMembersCount = initialClusterMembersCount;
+            this.config = config;
+
+            hazelcastInstances = factory.newInstances(config, initialClusterMembersCount);
+            HazelcastTestSupport.assertClusterSizeEventually(initialClusterMembersCount, hazelcastInstances[0]);
+            failingInstance = factory.newHazelcastInstance(config);
+            HazelcastTestSupport.assertClusterSizeEventually(initialClusterMembersCount + 1, hazelcastInstances[0]);
+            HazelcastTestSupport.waitAllForSafeState(hazelcastInstances);
+            HazelcastTestSupport.waitAllForSafeState(failingInstance);
+            client = null;
+        }
+
+        public HazelcastInstance createClient(ClientConfig clientConfig) {
+            client = factory.newHazelcastClient(clientConfig);
+            return client;
+        }
+
+        public void cleanUp() {
+            Arrays.stream(hazelcastInstances).forEach(instance -> instance.getLifecycleService().terminate());
+            client.getLifecycleService().terminate();
```

Comment:
minor: i believe it would be better to close the client first, so that we don't waste time trying to reconnect to the cluster

Reply:
Done